internal/adapters/gitlabtokenrefresher/gitlabtokenrefresher.go

```diff
@@ -0,0 +1,105 @@
+package gitlabtokenrefresher
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/SwissDataScienceCenter/renku-gateway-v2/internal/adapters/redisadapters"
+	"github.com/SwissDataScienceCenter/renku-gateway-v2/internal/models"
+	"github.com/go-co-op/gocron"
+)
+
+type tokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	Type         string `json:"token_type"`
+	ExpiresIn    int64  `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+	Scope        string `json:"scope"`
+	CreatedAt    int64  `json:"created_at"`
+}
+
+func ScheduleRefreshExpiringTokens(ctx context.Context, redisAdapter redisadapters.RedisAdapter, gitlabTokenRefreshURL string, clientID string, clientSecret string, minsToExpiration int) error {
+	// schedule token refresh evaluations
+	s := gocron.NewScheduler(time.UTC)
+	job, err := s.Every(minsToExpiration).Minutes().Do(refreshExpiringTokens, ctx, redisAdapter, gitlabTokenRefreshURL, clientID, clientSecret, minsToExpiration)
+	s.StartBlocking()
+	if err != nil {
+		log.Printf("Reading body failed: %s", err)
+	} else {
+		log.Printf("Job starting: %v", job)
+	}
+	return err
+}
+
+func refreshExpiringTokens(ctx context.Context, redisAdapter redisadapters.RedisAdapter, gitlabTokenRefreshURL string, clientID string, clientSecret string, minsToExpiration int) error {
```

Comment:
Define and use an interface here instead of `RedisAdapter`. You do not even need the full functionality that the Redis adapter offers.

All you need is the following as far as I can tell:
```
type RefresherTokenStore interface {
    GetRefreshToken(context.Context, string) (string, error)
    GetAccessToken(context.Context, string) (string, error)
    SetRefreshToken(context.Context, models.AccessToken) error
    SetAccessToken(context.Context, models.RefreshToken) error
}
```

```suggestion
func refreshExpiringTokens(ctx context.Context, tokenStore RefresherTokenStore, gitlabTokenRefreshURL string, clientID string, clientSecret string, minsToExpiration int) error {
```

With the interface you are truly saying "I do not care what storage is used the tokens as long as it can set and get tokens for me". With a concrete type you are much more limited.